### PR TITLE
[image-builder] Fix ignoring of user-passed authentication

### PR DESCRIPTION
**Pull request by geropl**
## Description
With https://github.com/gitpod-io/gitpod/pull/19474 a bug was introduced that ignores two registry authentication special cases:
 1. installation defined registry auth ([code](https://github.com/gitpod-io/gitpod/pull/19474/files#diff-b1c591e18af598ccba93b570ce7c97dece0988c9109a6d3be82bbe0df1fbd619L1057-L1062))
    - this is an old self-hosted feature, and we don't seem to be using it anymore :heavy_check_mark:  
 2. user-defined registry auth (the `GITPOD_IMAGE_AUTH` variable)
    - this one is fixed with this PR :point_down:  

## Related Issue(s)

Fixes ENT-72

## How to test
 - start an image build referencing a base image in a private repository (via `GITPOD_IMAGE_AUTH`)

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-72-fixab59193ca7</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-72-fixab59193ca7.preview.gitpod-dev.com/workspaces" target="_blank">gpl-72-fixab59193ca7.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - gpl-72-fix-image-auth-gha.25210</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-gpl-72-fixab59193ca7%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by geropl** [comment]
@iQQBot Could you review this PR, and see if it makes sense?

File: components/image-builder-mk3/pkg/orchestrator/orchestrator.go
Comment (kylos101) [inline components/image-builder-mk3/pkg/orchestrator/orchestrator.go:235]: TIL about BuildRegistryAuthSelective

Comment (kylos101) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:235]: Related: https://github.com/gitpod-io/gitpod/blob/aa58596eb4bc079da3c836dfef3eacf8f18e09d8/components/server/src/workspace/workspace-starter.ts#L1074-L1076

Comment (geropl) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:235]: BTW just pushed to change it to this:
```
		if req.Auth.GetSelective() != nil {
			// allow access to baseImage repository so we can look it up later
			req.Auth.GetSelective().AllowBaserep = true
			reqauth = o.AuthResolver.ResolveRequestAuth(req.Auth)
		}
```
to avoid the lurking nil deref :see_no_evil:

Comment (geropl) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:235]: Regarding https://github.com/gitpod-io/gitpod/pull/19745#discussion_r1601815771: this is the self-hosted feature I mentioned in the PR description:

> installation defined registry auth ([code](https://github.com/gitpod-io/gitpod/pull/19474/files#diff-b1c591e18af598ccba93b570ce7c97dece0988c9109a6d3be82bbe0df1fbd619L1057-L1062))
>  - this is an old self-hosted feature, and we don't seem to be using it anymore ✔️

Comment (kylos101) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:235]: I am not seeing the push.

Instead of force pushing, can you push a new commit? Makes it easier to see differences, we can squash later.

Comment (kylos101) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go:235]: req.Auth can be nil, too (perhaps less likely, but still)

**Review by kylos101 — APPROVED** [review]
Works well!

I tested the following:
1. Previously built workspace images that are dependent on private registries in a custom dockerfile are able to start workspaces w/o building or being dependent on having GITPOD_IMAGE_AUTH set on the repo.
2. A change to the private image in a custom dockerfile dependent on a private registry fails the build with a 502 if the GITPOD_IMAGE_AUTH is not set (image-builder-bob needs a credential).
3. Building on the previous test, once the GITPOD_IMAGE_AUTH is set in the repo, a workspace image is built, and a workspace started. It can also be restarted.
4. I can also use the same private image directly (so no custom docker file). It restarts fine, and if I delete the workspace-image it gets rebuilt.

I did not test private ECR. Let's plan to do against dogfood Thursday, and follow-up if needed? It's easier to land main-gha changes in Dedicated.

Before removing the hold, let's assert the problem can be recreated in a gen113 preview. That'll help us increase confidence that this resolves the issue. I'll do that now.

**Comment by geropl** [comment]
> Before removing the hold, let's assert the problem can be recreated in a gen113 preview. That'll help us increase confidence that this resolves the issue. I'll do that now.

Awesome, thank you for your thorough testing :orange_heart:

File: components/image-builder-mk3/pkg/orchestrator/orchestrator.go
Comment (geropl) [reply components/image-builder-mk3/pkg/orchestrator/orchestrator.go]: Done

**Comment by geropl** [comment]
Whoa, blocked by needing :heavy_check_mark: from an old EXP team member... :grimacing:

**Review by AlexTugarev — APPROVED** [review]

**Comment by geropl** [comment]
Thx @AlexTugarev for the unblock :bow:

**Pull request merged**
